### ISSUE TITLE
Missing tooth decoder

### ISSUE THREAD

**Issue by larmicfre** [via/tfi-computer#88]
One of the most popular aftermaket options which provides very good resolution. A specified number of equally spaced teeth with one or more missing teeth. Typical patterns, for example, used by many are 36-1 and 60-2.

**Comment by via** [comment]
I completely agree.  There are a few things that will need to happen before this can be implemented:
1) Most importantly, something to test with.  I'll probably eventually switch my main vehicle to a missing tooth crank wheel, at which point this requierment is met.  Alternatively, if anyone else has a suitable test vehicle, that could work too.  I can use a wheel on a drill for experimentation.

2) As we discussed, I think performance is an issue.  Right now we use each tooth as a trigger to both decode position and recalculate all fueling and ignition.  I think that would be an issue for a 60-2 wheel.  The DMA wheels work last year was part of the work to get this better -- we can take in precise times of tooth events and then handle them in batches, allowing us to soft-degrade at higher rpms (effectively transitioning to redoing calcs every other tooth or so as necessary).  However, the right way to solve this is probably to a) speed up decode and calcs, and b) more intelligently choose which events need rescheduling.  Right now we reschedule *all* (potentially 24) events on each tooth. I think only doing immediate recalculation of the immediate events (an ongoing or upcoming ignition/fuel only) is sufficient, and we can use the that 'soft-degrade' for the later ones.  I need to collect these thoughts in another issue.

3) All of the work for (2) will become much easier if I rip out the tire fire of the stats code and replace it with something more usable.  We need to know realtime how much decode/fuel compute capacity there is, and how long things are taking, outside of me testing on a bench.  This should also be collected into a new issue.

**Comment by via** [comment]
Re: (2), I started some benchmarking code that might eventually get into master.

Master currently outputs figures fairly consistently looking like:
```
calculate_fueling: 3397 ns
schedule_event: 4246 ns
```

Naively, that sounds to me like that with a max of 24 events, scheduling events could all happen within 100 uS, and 6000 rpms with a 60-2 wheel gives a 277 uS budget, so it should be great.  For whatever reason though, if I actually measure how long the decode ISR (decode + fuel calc + schedule X events) takes with a toggled pin and an analyzer, its usually about 180 uS, which is nothing for my 24 tooth cam wheel, but will become an issue around 4000 rpms for a 60-1 crank wheel.

However, something as simple as re-enabling LTO, which I removed just for simplicity, produces:
```
calculate_fueling: 1577 ns
schedule_event: 2210 ns       
```

Much faster! It doesn't 100% translate into a 2x performance gain.  It brings the ISR time down to 120-130 uS.  Still, this would be enough to run a 60-2 wheel at 6000 rpms with no code changes.

Going forward, aside from getting the benchmark code mergable and maybe some automated runs as master progresses, I need to figure out whats causing those ISRs to take so much longer than what I'd expect from 24 4ish uS schedules.  The primary suspects are the amount of time spent in interrupts during these calculations for
 - Buffer swaps -- I know this takes about 25% of the total cpu time
 - Sensor calcs -- These aren't cheap either, I need benchmarking here too

I might branch off this investigation into a separate issue, since more benchmarking and profiling is needed. However, I'm less worried about this now -- a 36-1 wheel on the crank is super feasible with the current code, and a 60-2 wheel is pretty close for most use cases.

**Comment by via** [comment]
After a bit more debugging, it turns out the current stats framework is a huge performance hog. Disabling that, keeping LTO, running the 24 cam teeth at 30k rpms was about a 50% cpu usage, which implies to me a 60-2 crank wheel at 10k rpms would be perfectly fine.

There's a metrics revamp in a branch that depends on completing the new cbor logging, which is feature complete and will be merged pending a new UI that works with it.

**Comment by via** [comment]
This is implemented now. I'll move the performance discussion to another issue down the road, though I think both wheel types are not a performance issue anymore.
